### PR TITLE
Fix issues when LIB recorded from previous shutdown is at the edge of a gap

### LINKED ISSUE #209: The start from LIB logic cannot recover from shutdown when LIB happen to be the first on after a gap
Currently the LIB used for recover is calculated as the height calculated from the EOS LIB timestamp **minus 1** to make sure it's not a partial block.

It will usually work except when the EOS LIB is just hitting the end of a gap, in such case the calculated evm block will be empty ones with 0x0 mixHash and this will cause panic.

Not urgent one as the case is rare and we have flag to manually start from earlier blocks.

### PR DESCRIPTION
**Pull request by yarkinwho**
Resolves #209

### CODE REVIEW

File: src/engine_plugin.cpp [hunk 1]
```diff
@@ -135,6 +135,28 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
...
+         // The search is not that optimized. However, this function will only be called during initialization process
+         // in rara cases and the gap in eos will not be too large in most cases. Therefore, it should be fine to leave 
+         // the search process simple.
+         SILK_INFO << "Search for block containing a valid eos id. Start from:" << "#" << start_height;
+         silkworm::db::ROTxn txn(db_env);
+         do {
+            auto res = silkworm::db::read_canonical_header(txn, start_height);
+            if(!res) return {};
+
+            for (int i = 0; i < 32; ++i) {
```
Comment (elmato) [inline src/engine_plugin.cpp:149]: You can use `!is_zero(res->prev_randao)`

File: src/engine_plugin.cpp [hunk 2]
```diff
@@ -135,6 +135,28 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
...
       }
 
+      std::optional<silkworm::BlockHeader> find_block_with_valid_eos_id(uint64_t start_height) {
+         // Search for the last block with a valid eos id in it.
+         // The search is not that optimized. However, this function will only be called during initialization process
+         // in rara cases and the gap in eos will not be too large in most cases. Therefore, it should be fine to leave 
+         // the search process simple.
+         SILK_INFO << "Search for block containing a valid eos id. Start from:" << "#" << start_height;
+         silkworm::db::ROTxn txn(db_env);
+         do {
```
Comment (elmato) [inline src/engine_plugin.cpp:145]: why not just use `while` and decrement `start_height` inside the while body?

File: src/engine_plugin.cpp [hunk 3]
```diff
@@ -135,6 +135,28 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
...
+               if (res->prev_randao.bytes[i]) {
+                  SILK_INFO << "Found block at: " << "#" << start_height;
+                  return res;
+               }
+            }
+         }
+         while(--start_height > 0);
+         return {};
+      }
+
```
Comment (elmato) [inline src/engine_plugin.cpp:159]: // Alternative implementation
```c++
std::optional<silkworm::BlockHeader> res;
while(start_height>0) {
  auto res = silkworm::db::read_canonical_header(txn, start_height);
  if(res && !is_zero(res->prev_randao)) {
     break;
  }
 --start_height;
}
return res;

Comment (yarkinwho) [reply src/engine_plugin.cpp:155]: fixed as this

Comment (yarkinwho) [reply src/engine_plugin.cpp:154]: Refactor to do while again to handle the genesis case properly.

**Review by elmato — APPROVED** [review]

**Pull request merged**